src/tactic/explode_widget.lean

```diff
@@ -0,0 +1,305 @@
+                     return $ [h "td" [cn "ba bg-dark-green tc"] "Rule", 
+                               h "td" [cn "ba tc"] t]
+| t := return $ [h "td" [cn "ba bg-dark-green tc"] "Rule", 
+                 h "td" [cn "ba tc"] t.to_string]
+
+/--
+Render a row that contains the sub-proofs, i.e., the proofs of the
+arguments.
+-/
+meta def proof_row {γ} (args : list (html γ)): list (html γ) := 
+[h "td" [cn "ba bg-dark-green tc"] "Proofs", h "td" [cn "ba tc"] 
+    [h "details" [] $
+        (h "summary" 
+            [attr.style [("color", "orange")]] 
+                "Details")::args]
+]
+
+/--
+Combine the goal row, id row, rule row and proof row to make them a table.
+-/
+meta def assemble_table {γ} (gr ir rr) : list (html γ) → html γ
+| [] := 
+h "table" [cn "collapse"] 
+    [h "tbody" []
+        [h "tr" [] gr, h "tr" [] ir, h "tr" [] rr]
+    ]
+| pr := 
+h "table" [cn "collapse"] 
+    [h "tbody" []
+        [h "tr" [] gr, h "tr" [] ir, h "tr" [] rr, h "tr" [] pr]
+    ]
+
+/--
+Render a table for a given entry.
+-/
+meta def assemble (es : entries): entry → tactic (html empty)
+| ⟨e, l, d, status.sintro, t, ref⟩ := do
+    gr ← goal_row e, ir ← id_row l, rr ← rule_row $ thm.string "Assumption",
+    return $ assemble_table gr ir rr []
+| ⟨e, l, d, status.intro, t, ref⟩ := do
+    gr ← goal_row e, ir ← id_row l, rr ← rule_row $ thm.string  "Assumption",
+    return $ assemble_table gr ir rr []
+| ⟨e, l, d, st, t, ref⟩ := do 
+    gr ← goal_row e ff, ir ← id_row l, rr ← rule_row t,
+    let el : list entry := list.map (lookup_lines es) ref, 
+    ls ← monad.mapm assemble el,
+    let pr := proof_row $ ls.intersperse (h "br" [] []),
+    return $ assemble_table gr ir rr pr
+
+/--
+Render a widget from given entries.
+-/
+meta def explode_component (es : entries) : tactic (html empty) := 
+let concl := lookup_lines es (es.l.length - 1) in assemble es concl
+
+/--
+Explode a theorem and return entries.
+-/
+meta def explode_entries (n : name) (hide_non_prop := tt) : tactic entries :=
+do expr.const n _ ← resolve_name n | fail "cannot resolve name",
+  d ← get_decl n,
+  v ← match d with
+  | (declaration.defn _ _ _ v _ _) := return v
+  | (declaration.thm _ _ _ v)      := return v.get
+  | _                  := fail "not a definition"
+  end,
+  t ← pp d.type,
+  explode_expr v hide_non_prop
+
+end explode_widget
+
+open lean lean.parser interactive explode_widget
+
+/--
+User command of the explode widget.
+-/
+@[user_command]
+meta def explode_widget_cmd (_ : parse $ tk "#explode_widget") : lean.parser unit :=
+do ⟨li,co⟩ ← cur_pos,
+    n ← ident,
+    es ← explode_entries n,
+    comp ← parser.of_tactic (do html ← explode_component es,
+    c ← pure $ component.stateless (λ _, [html]),
+    pure $ component.ignore_props $ component.ignore_action $ c),
+    save_widget ⟨li, co - "#explode_widget".length - 1⟩ comp,
+    trace "successfully rendered widget",
+    skip
+    .
+
+end tactic
```

Comment:
This suggestion is to make `#explode_widget` show up in our tactic / user command docs, see [the doc entry for `#explode`](https://leanprover-community.github.io/mathlib_docs/commands.html##explode).

There are two options. First, we could create a combined doc entry for both `#explode` and `#explode_widget`. To do that, delete the `add_tactic_doc_entry` that appears after `tactic.explode_cmd` and add a sentence or two mentioning `#explode_widget` to the docstring of `tactic.explode_cmd`.
```suggestion

add_tactic_doc
{ name       := "#explode / #explode_widget",
  category   := doc_category.cmd,
  decl_names := [`tactic.explode_cmd, `tactic.explode_widget_cmd],
  inherit_description_from := `tactic.explode_cmd,
  tags       := ["proof display", "widgets"] }

end tactic
```

The alternative is to keep the tactic doc entry for `#explode` and have a separate one for `#explode_widget`. If you prefer to do that, then you should expand the doc string for `explode_widget_cmd` and commit the following instead (it'd also be good to mention `#explode_widget` in the doc string of `explode_cmd`):
```suggestion

add_tactic_doc
{ name       := "#explode_widget",
  category   := doc_category.cmd,
  decl_names := [`tactic.explode_widget_cmd],
  tags       := ["proof display", "widgets"] }

end tactic
```

Reply:
The first option seems more reasonable. I guess in the end we only want one of them -- the ASCII art version can also be part of the widget.